### DISCUSSION TITLE
Link on delegate inside of anonymous method

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#2260] [Language Ideas]
@vsfeedback commented on [Fri Feb 22 2019](https://github.com/dotnet/roslyn/issues/33619)

<p>I suggest to add &quot;this&quot;-analog for anonymous methods for easier use when it is essential to remove handler from event inside of handler itself<br>before</p><pre>Timer timer = new Timer();
//ignoring right signature for simplifying
Action a = null;
a = () =&gt;
{
    //Dosmth
    timer.Elapsed -= a;
}
timer.Elapsed += a;</pre>
after:<br><br><pre>Timer timer = new Timer();
//ignoring right signature for simplifying
timer.Elapsed += () =&gt;
{
    //DoSmth&lt;br&gt;   //thisDelegeate is the link on current delegate. Exactly this i suggest to add&lt;br&gt;    timer.Elapsed -= thisDelegeate;
}</pre>
<br>

_This issue has been moved from https://developercommunity.visualstudio.com/content/idea/460725/link-on-delegate-inside-of-anonymous-method.html
VSTS ticketId: 795742_
_These are the original issue comments:_

Jane Wu [MSFT] on 2/18/2019, 11:58 PM (3 days ago): <p>Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.&nbsp;We'll provide an update once the issue has been triaged by the product team.</p>

---

@gafter commented on [Sat Feb 23 2019](https://github.com/dotnet/roslyn/issues/33619#issuecomment-466692903)

This is a language request.  Moving to csharplang

**Comment by Joe4evr** [comment]
@gafter Bit off-topic, but can you start using this instead to move threads in-place?
![](https://cdn.discordapp.com/attachments/372892461137133568/543192818055905290/unknown.png)

**Comment by SilentSin** [comment]
Rather than needing a new keyword, this feature could probably reuse the `delegate` keyword.
```
timer.Elapsed += () =>
{
    //DoSmth
    timer.Elapsed -= delegate;
}
```

**Comment by gafter** [comment]
> can you start using this instead to move threads in-place

No, because I am not an admin for these repositories.

**Comment by kharai-tor** [comment]
The `delegate` keyword might not be a good idea. That's a lambda, not a delegate and even though there's an implicit conversion between the two, it's still incorrect and therefore confusing.

**Comment by jnm2** [comment]
@Joe4evr / @gafter

Please send feedback to GitHub about this. I haven't been able to use the issue transfer feature yet at NUnit either for the same reason, and it would be really useful there.

https://twitter.com/jnm236/status/1098052245894193152
